sdk/storage/azure-storage-blob-cryptography/src/main/java/com/azure/storage/blob/specialized/cryptography/BlobDecryptionPolicy.java

```diff
@@ -0,0 +1,485 @@
+                        beginningOfEndAdjustment = encryptedBlobRange.offsetAdjustment()
+                            + encryptedBlobRange.originalRange().getCount();
+                    }
+
+                    /*
+                    The end of the Cipher output lies after customer requested data (in the end adjustment) and
+                    should be trimmed.
+                     */
+                    if (decryptedBytes + totalOutputBytes.longValue() > beginningOfEndAdjustment) {
+                        long amountPastEnd // past the end of user-requested data.
+                            = decryptedBytes + totalOutputBytes.longValue() - beginningOfEndAdjustment;
+                        /*
+                        Note that amountPastEnd can only be up to 16, so the cast is safe. We do not need to worry
+                        about limit() throwing because we allocated at least enough space for decryptedBytes and
+                        the newLimit will be less than that. In the case where this Cipher output starts after the
+                        beginning of the endAdjustment, we don't want to send anything back, so we set limit to be
+                        the same as position.
+                         */
+                        int newLimit = totalOutputBytes.longValue() <= beginningOfEndAdjustment
+                            ? decryptedBytes - (int) amountPastEnd : plaintextByteBuffer.position();
+                        plaintextByteBuffer.limit(newLimit);
+                    } else if (decryptedBytes + totalOutputBytes.longValue()
+                        > encryptedBlobRange.offsetAdjustment()) {
+                     /*
+                    The end of this Cipher output is before the end adjustment and after the offset adjustment, so
+                    it will lie somewhere in customer requested data. It is possible we allocated a ByteBuffer that
+                    is slightly too large, so we set the limit equal to exactly the amount we decrypted to be safe.
+                     */
+                        plaintextByteBuffer.limit(decryptedBytes);
+                    } else {
+                     /*
+                    Else: The end of this ByteBuffer will not reach the beginning of customer-requested data. Make
+                    it effectively empty.
+                     */
+                        plaintextByteBuffer.limit(plaintextByteBuffer.position());
+                    }
+
+                    totalOutputBytes.addAndGet(decryptedBytes);
+                    return plaintextByteBuffer;
+                });
+            });
+    }
+
+    /**
+     * Gets and validates {@link EncryptionData} from a Blob's metadata
+     *
+     * @param encryptedDataString {@code String} of encrypted metadata
+     *
+     * @return {@link EncryptionData}
+     */
+    private EncryptionData getAndValidateEncryptionData(String encryptedDataString) {
+        if (encryptedDataString == null) {
```

Comment:
This should throw NPE instead - https://azure.github.io/azure-sdk/java_implementation.html#java-errors-system-errors

Reply:
This is actually by design. The guidelines state throw a NPE when null is unexpected, but null is a valid value here, it just means that there's nothing to parse. We use that in decrypt blob to indicate skipping the decryption step and go right to trimming. 

---
In reply to: [332684738](https://github.com/Azure/azure-sdk-for-java/pull/5528#discussion_r332684738) [](ancestors = 332684738)